### ISSUE TITLE
gv_const_sv broken in blead, suspect is 2eaf799e

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14214]
Migrated from [rt.perl.org#123145](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123145) (status was 'resolved')

Searchable as RT123145$

**Comment by p5pRT** [comment]
### From sog@msg.mx


This is a bug report for perl from sog@&#8203;msg\.com\.mx\,
generated with the help of perlbug 1\.40 running under perl 5\.21\.6\.


Not used in core but gv\_const\_sv is documented to return NULL unless the passed
GV references an inlineable subroutine\, but after FC's commit 2eaf799e
"Avoid creating GVs when subs are declared" it returns the CV unless the sub is
declared with an explicit package\.

I think that gv\_const\_sv should be fixed\, and the changes in the stash content
prominently documented\.

Compare what is produced by the attached test in  \<= 5\.20\.1 vs blead\.

\-\-
sortiz

======
\#\!/usr/bin/perl

use Inline C => \<\<'END\_OF\_CODE';

void gv\_const\_test\(SV \*name\) \{
&nbsp;    GV \*gv;
&nbsp;    if\(SvTYPE\(name\) == SVt\_PV\) \{
&nbsp;    HV \*stash = gv\_stashpv\("main"\,0\);
&nbsp;    HE \*he = hv\_fetch\_ent\(stash\, name\, 0\, 0\);
&nbsp;    gv = \(GV \*\)HeVAL\(he\);
&nbsp;    \}
&nbsp;    else gv = \(GV \*\)name;
&nbsp;    if\(SvROK\(gv\)\) warn\("Glob %s reftype&#8203;: %d\\n"\,SvPV\_nolen\(gv\)\,SvTYPE\(SvRV\(gv\)\)\);
&nbsp;    SV \*sv = gv\_const\_sv\(gv\);
&nbsp;    if\(sv\) \{
&nbsp;    warn\("%s is constant&#8203;: %s\\n"\, SvPV\_nolen\(name\)\, SvPV\_nolen\(sv\)\);
&nbsp;    \} else \{
&nbsp;    warn\("%s isn't constant\.\\n"\, SvPV\_nolen\(name\)\);
&nbsp;    \}
\}

END\_OF\_CODE

sub one \(\) \{ 1; \}
sub baz \(\) \{ "baz"; \}
sub meth \{ my $self = shift; return $self; \}
sub main&#8203;::foo \{ \};

gv\_const\_test\($\_\) for qw\(one baz meth foo\);
gv\_const\_test\($&#8203;::\{$\_\}\) for qw\(one baz meth foo\);
gv\_const\_test\(\*\{$\_\}\) for qw\(one baz meth foo\);

\_\_END\_\_


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl 5\.21\.6&#8203;:

Configured by sog at Wed Nov  5 19&#8203;:06&#8203;:14 CST 2014\.

Summary of my perl5 \(revision 5 version 21 subversion 6\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: 6b6b21da8e7a4aa561c0b61f5a38769acc035b26
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.16\.6\-203\.fc20\.x86\_64\, archname=x86\_64\-linux
&nbsp;    uname='linux monica 3\.16\.6\-203\.fc20\.x86\_64 \#1 smp sat oct 25 12&#8203;:44&#8203;:32 utc 2014 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-de \-Dprefix=/work/common/perl5/perls/perl\-blead \-Dusedevel \-Aeval&#8203;:scriptdir=/work/common/perl5/perls/perl\-blead/bin'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-D\_FORTIFY\_SOURCE=2'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.3 20140911 \(Red Hat 4\.8\.3\-7\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\, doublekind=3
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16\, longdblkind=3
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib /lib/\.\./lib64 /usr/lib/\.\./lib64 /lib /lib64 /usr/lib64 /usr/local/lib64
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.18\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.18'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector\-strong'




---
@&#8203;INC for perl 5\.21\.6&#8203;:
&nbsp;    blib/lib
&nbsp;    blib/arch
&nbsp;    /work/common/perl5/perls/perl\-blead/lib/site\_perl/5\.21\.6/x86\_64\-linux
&nbsp;    /work/common/perl5/perls/perl\-blead/lib/site\_perl/5\.21\.6
&nbsp;    /work/common/perl5/perls/perl\-blead/lib/5\.21\.6/x86\_64\-linux
&nbsp;    /work/common/perl5/perls/perl\-blead/lib/5\.21\.6
&nbsp;    \.



---
Environment for perl 5\.21\.6&#8203;:
&nbsp;    HOME=/home/sog
&nbsp;    LANG=en\_US\.utf8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/work/common/perl5/bin&#8203;:/work/common/perl5/perls/perl\-blead/bin&#8203;:/opt/google\-cloud\-sdk/bin&#8203;:/opt/google\-cloud\-sdk/bin&#8203;:/usr/lib64/qt\-3\.3/bin&#8203;:/opt/google\-cloud\-sdk/bin&#8203;:/usr/lib64/ccache&#8203;:/usr/libexec/lightdm&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/local/sbin&#8203;:/usr/sbin&#8203;:/home/sog/bin&#8203;:/opt/android\-sdk\-linux/tools&#8203;:/opt/android\-sdk\-linux/platform\-tools
&nbsp;    PERL5LIB=blib/lib&#8203;:blib/arch
&nbsp;    PERLBREW\_BASHRC\_VERSION=0\.71
&nbsp;    PERLBREW\_CPAN\_MIRROR=http&#8203;://www\.msg\.com\.mx/CPAN
&nbsp;    PERLBREW\_HOME=/home/sog/\.perlbrew
&nbsp;    PERLBREW\_MANPATH=/work/common/perl5/perls/perl\-blead/man
&nbsp;    PERLBREW\_PATH=/work/common/perl5/bin&#8203;:/work/common/perl5/perls/perl\-blead/bin
&nbsp;    PERLBREW\_PERL=perl\-blead
&nbsp;    PERLBREW\_ROOT=/work/common/perl5
&nbsp;    PERLBREW\_VERSION=0\.71
&nbsp;    PERLDOC\_PAGER=less \-R
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_DESTRUCT\_LEVEL=2
&nbsp;    PERL\_DL\_NONLAZY=1
&nbsp;    PERL\_UNICODE=SDAL
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu Nov 06 16&#8203;:22&#8203;:29 2014\, sortiz wrote&#8203;:
> 
> This is a bug report for perl from sog@&#8203;msg\.com\.mx\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.6\.
> 
> 
> Not used in core but gv\_const\_sv is documented to return NULL unless
> the passed
> GV references an inlineable subroutine\, but after FC's commit 2eaf799e
> "Avoid creating GVs when subs are declared" it returns the CV unless
> the sub is
> declared with an explicit package\.
> 
> I think that gv\_const\_sv should be fixed\, and the changes in the stash
> content
> prominently documented\.
> 
> Compare what is produced by the attached test in  \<= 5\.20\.1 vs blead\.
> 
> \-\-
> sortiz
> 
> ======
> \#\!/usr/bin/perl
> 
> use Inline C => \<\<'END\_OF\_CODE';
> 
> void gv\_const\_test\(SV \*name\) \{
>     GV \*gv;
>     if\(SvTYPE\(name\) == SVt\_PV\) \{
>         HV \*stash = gv\_stashpv\("main"\,0\);
>         HE \*he = hv\_fetch\_ent\(stash\, name\, 0\, 0\);
>         gv = \(GV \*\)HeVAL\(he\);
>     \}
>     else gv = \(GV \*\)name;
>     if\(SvROK\(gv\)\) warn\("Glob %s reftype&#8203;:
> %d\\n"\,SvPV\_nolen\(gv\)\,SvTYPE\(SvRV\(gv\)\)\);
>     SV \*sv = gv\_const\_sv\(gv\);
>     if\(sv\) \{
>         warn\("%s is constant&#8203;: %s\\n"\, SvPV\_nolen\(name\)\,
> SvPV\_nolen\(sv\)\);
>     \} else \{
>         warn\("%s isn't constant\.\\n"\, SvPV\_nolen\(name\)\);
>     \}
> \}
> 
> END\_OF\_CODE
> 
> sub one \(\) \{ 1; \}
> sub baz \(\) \{ "baz"; \}
> sub meth \{ my $self = shift; return $self; \}
> sub main&#8203;::foo \{ \};
> 
> gv\_const\_test\($\_\) for qw\(one baz meth foo\);
> gv\_const\_test\($&#8203;::\{$\_\}\) for qw\(one baz meth foo\);
> gv\_const\_test\(\*\{$\_\}\) for qw\(one baz meth foo\);
> 
> \_\_END\_\_

Data from running the OP's program on perl\-5\.20\.1 and on blead&#8203;:

\#\#\#\#\#
$ perl \-v | head \-2 | tail \-1
This is perl 5\, version 20\, subversion 1 \(v5\.20\.1\) built for x86\_64\-linux

$ perl 123145\.pl 
one is constant&#8203;: 1
baz is constant&#8203;: baz
meth isn't constant\.
foo isn't constant\.
\*main&#8203;::one is constant&#8203;: 1
\*main&#8203;::baz is constant&#8203;: baz
\*main&#8203;::meth isn't constant\.
\*main&#8203;::foo isn't constant\.
\*main&#8203;::one is constant&#8203;: 1
\*main&#8203;::baz is constant&#8203;: baz
\*main&#8203;::meth isn't constant\.
\*main&#8203;::foo isn't constant\.
\#\#\#\#\#
$ \./bin/perl \-v | head \-2 | tail \-1

This is perl 5\, version 21\, subversion 6 \(v5\.21\.6 \(v5\.21\.5\-462\-g0d71ffa\)\) built for x86\_64\-linux

$ \./bin/perl \-Ilib /home/jkeenan/learn/perl/p5p/123145\.pl 
Glob SCALAR\(0xf97dd8\) reftype&#8203;: 1
one is constant&#8203;: 1
Glob SCALAR\(0xf97da8\) reftype&#8203;: 3
baz is constant&#8203;: baz
Glob CODE\(0xf97d78\) reftype&#8203;: 13
meth is constant&#8203;: 
foo isn't constant\.
Glob SCALAR\(0xf97dd8\) reftype&#8203;: 5
SCALAR\(0xf97dd8\) is constant&#8203;: 1
Glob SCALAR\(0xf97da8\) reftype&#8203;: 3
SCALAR\(0xf97da8\) is constant&#8203;: baz
Glob CODE\(0xf97d78\) reftype&#8203;: 13
CODE\(0xf97d78\) is constant&#8203;: 
\*main&#8203;::foo isn't constant\.
\*main&#8203;::one is constant&#8203;: 1
\*main&#8203;::baz is constant&#8203;: baz
\*main&#8203;::meth isn't constant\.
\*main&#8203;::foo isn't constant\.
\#\#\#\#\#

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Nov 06 16&#8203;:22&#8203;:29 2014\, sortiz wrote&#8203;:
> Not used in core but gv\_const\_sv is documented to return NULL unless
> the passed
> GV references an inlineable subroutine\, but after FC's commit 2eaf799e
> "Avoid creating GVs when subs are declared" it returns the CV unless
> the sub is
> declared with an explicit package\.
> 
> I think that gv\_const\_sv should be fixed\, and the changes in the stash
> content
> prominently documented\.

Please try the attached fix for gv\_const\_sv\(\)\.

I think the documentation change is needed too\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-123145-gv_const_sv-handle-subrefs-directly-in-t.patch</summary>

```diff
From 69171862a50d5659c72391dbabf3003cb111483e Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 2 Mar 2015 15:59:32 +1100
Subject: [PATCH] [perl 123145] gv_const_sv() handle subrefs directly in the
 stash

With 2eaf799e sub declarations can store sub references directly in
the stash, but gv_const_sv(), which might be passed an entry from
the stash didn't handle that case.
---
 MANIFEST                       |    1 +
 ext/XS-APItest/APItest.pm      |    2 +-
 ext/XS-APItest/APItest.xs      |   20 ++++++++++++++++++++
 ext/XS-APItest/t/gv_const_sv.t |   18 ++++++++++++++++++
 gv.c                           |    2 +-
 5 files changed, 41 insertions(+), 2 deletions(-)
 create mode 100644 ext/XS-APItest/t/gv_const_sv.t

diff --git a/MANIFEST b/MANIFEST
index aad1be4..2595c39 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -3996,6 +3996,7 @@ ext/XS-APItest/t/fetch_pad_names.t	Tests for UTF8 names in pad
 ext/XS-APItest/t/gotosub.t	XS::APItest: tests goto &xsub and hints
 ext/XS-APItest/t/grok.t		XS::APItest: tests for grok* functions
 ext/XS-APItest/t/gv_autoload4.t	XS::APItest: tests for gv_autoload4() and variants
+ext/XS-APItest/t/gv_const_sv.t	XS::APItest: test gv_const_sv()
 ext/XS-APItest/t/gv_fetchmeth_autoload.t	XS::APItest: tests for gv_fetchmeth_autoload() and variants
 ext/XS-APItest/t/gv_fetchmethod_flags.t	XS::APItest: tests for gv_fetchmethod_flags() and variants
 ext/XS-APItest/t/gv_fetchmeth.t		XS::APItest: tests for gv_fetchmeth() and variants
diff --git a/ext/XS-APItest/APItest.pm b/ext/XS-APItest/APItest.pm
index a759492..e4b7156 100644
--- a/ext/XS-APItest/APItest.pm
+++ b/ext/XS-APItest/APItest.pm
@@ -5,7 +5,7 @@ use strict;
 use warnings;
 use Carp;
 
-our $VERSION = '0.70';
+our $VERSION = '0.71';
 
 require XSLoader;
 
diff --git a/ext/XS-APItest/APItest.xs b/ext/XS-APItest/APItest.xs
index 2b7b349..68c164b 100644
--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -2375,6 +2375,26 @@ gv_autoload_type(stash, methname, type, method)
         }
 	XPUSHs( gv ? (SV*)gv : &PL_sv_undef);
 
+SV *
+gv_const_sv(SV *name)
+    PREINIT:
+        GV *gv;
+    CODE:
+        if (SvPOK(name)) {
+	    HV *stash = gv_stashpv("main",0);
+	    HE *he = hv_fetch_ent(stash, name, 0, 0);
+	    gv = (GV *)HeVAL(he);
+        }
+	else {
+	    gv = (GV *)name;
+        }
+        RETVAL = gv_const_sv(gv);
+        if (!RETVAL)
+            XSRETURN_EMPTY;
+	RETVAL = newSVsv(RETVAL);
+    OUTPUT:
+        RETVAL
+
 void
 whichsig_type(namesv, type)
     SV* namesv
diff --git a/ext/XS-APItest/t/gv_const_sv.t b/ext/XS-APItest/t/gv_const_sv.t
new file mode 100644
index 0000000..688fd48
--- /dev/null
+++ b/ext/XS-APItest/t/gv_const_sv.t
@@ -0,0 +1,18 @@
+#!perl
+
+use strict;
+use warnings;
+use Test::More tests => 6;
+
+use XS::APItest;
+
+sub foo () { "abc" }
+
+sub bar { }
+
+is(XS::APItest::gv_const_sv(*foo), "abc", "on const glob");
+is(XS::APItest::gv_const_sv("foo"), "abc", "on const by name");
+is(XS::APItest::gv_const_sv($::{"foo"}), "abc", "on const by lookup");
+is(XS::APItest::gv_const_sv(*bar), undef, "on non-const glob");
+is(XS::APItest::gv_const_sv("bar"), undef, "on non-const by name");
+is(XS::APItest::gv_const_sv($::{"bar"}), undef, "on non-const by lookup");
diff --git a/gv.c b/gv.c
index 41cebeb..e65b2d6 100644
--- a/gv.c
+++ b/gv.c
@@ -155,7 +155,7 @@ Perl_gv_const_sv(pTHX_ GV *gv)
 
     if (SvTYPE(gv) == SVt_PVGV)
 	return cv_const_sv(GvCVu(gv));
-    return SvROK(gv) && SvTYPE(SvRV(gv)) != SVt_PVAV ? SvRV(gv) : NULL;
+    return SvROK(gv) && SvTYPE(SvRV(gv)) != SVt_PVAV && SvTYPE(SvRV(gv)) != SVt_PVCV ? SvRV(gv) : NULL;
 }
 
 GP *
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Mar 01 21&#8203;:01&#8203;:19 2015\, tonyc wrote&#8203;:
> On Thu Nov 06 16&#8203;:22&#8203;:29 2014\, sortiz wrote&#8203;:
> > Not used in core but gv\_const\_sv is documented to return NULL unless
> > the passed
> > GV references an inlineable subroutine\, but after FC's commit 2eaf799e
> > "Avoid creating GVs when subs are declared" it returns the CV unless
> > the sub is
> > declared with an explicit package\.
> > 
> > I think that gv\_const\_sv should be fixed\, and the changes in the stash
> > content
> > prominently documented\.
> 
> Please try the attached fix for gv\_const\_sv\(\)\.

It looks good to me\, so I have applied it as 16a6e5a4fcd\.

> I think the documentation change is needed too\.

Do we document anywhere the use of IVs and PVs for sub stubs and refs for constants?  If so\, we should add it to that\.  But I don’t believe we do\.  In perlmod\, we warn against making any assumptions about stash entries that are not typeglobs\, which I think is sufficient\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
